File: projects/proj01_sol/code/src/pam_io.c
/*
* Authors:
*   Russell J. Hewett (rhewett@vt.edu)
*
* Not licensed for external distribution.
*/

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "image_types.h"
#include "image_memory.h"

#include "pam_io.h"


/*
* PGM format image loader for 8 bit gray scale images.
*
* Based on https://stackoverflow.com/questions/2693631/
* This is quite fragile, it does not support comments or other spacings in the
* PGM header.  Be careful using this!
*
* Note: This function acts as an allocator for image, as you cannot know the
*       size in advance!
*
* Arguments:
*   filename: name of file to load
*   image: Image struct to load in to
*
* Returns:
*   error code, 0 for success, 1 for failure
*/

int load_pgm_MImage8(char* filename, MImage8* image) {

    FILE* f;
    char format[2];
    int w, h, color_depth;
    int n_pixels;

    // Open the file
    f = fopen(filename, "rb");
    if (f == NULL) {
        fprintf(stderr, "Error opening file %s for read.\n", filename);
        return 1;
    }

    // // Read format, width, height, color depth
    if(fscanf(f, "%c%c %d %d %d\n", &format[0], &format[1], &w, &h, &color_depth) != 5){
        fprintf(stderr, "Error opening file %s: malformed pgm file.\n", filename);
        return 1;
    }

    // Fail if wrong format
    if(format[0] != 'P' || format[1] != '5') {
        fprintf(stderr, "Error opening file %s: Incorrect file format %c%c, must be P5.\n", 
                filename, format[0], format[1]);
        return 1;
    }

    if(color_depth > 255){
        fprintf(stderr, "Error opening file %s: Incorrect color depth %d, must be <255.\n", 
                filename, color_depth);
        return 1;
    }

    n_pixels = w*h;

    allocate_MImage8(image, w, h);
    initialize_MImage8(image);

    if(fread(image->data, sizeof(unsigned char), n_pixels, f) != n_pixels){
        fprintf(stderr, "Error opening file %s: malformed pgm file.\n", filename);
        return 1;
    }

    fclose(f);

    return 0;
}


/*
* PGM format image saver for 8 bit gray scale images.
*
* Based on code from Stephen Timmel.
*
* Arguments:
*   filename: name of file to save
*   image: Image struct to save
*
* Returns:
*   error code, 0 for success, 1 for failure
*/

int save_pgm_MImage8(char* filename, MImage8* image) {

    FILE* f;
    int n_pixels;

    // Open the file
    f = fopen(filename, "wb");
    if (f == NULL) {
        fprintf(stderr, "Error opening file %s for write.\n", filename);
        return 1;
    }

    // Write the ppm header, max color depth = 255
    fprintf(f, "P5 %d %d %d\n", image->width, image->height, 255);

    n_pixels = image->width*image->height;

    // Write entire image at once, as it is planar format
    fwrite(image->data, sizeof(unsigned char), n_pixels, f);

    fclose(f);

    return 0;
}


/*
* PPM format image saver for 24 bit RGB images.
*
* Based on code from Stephen Timmel.
*
* Arguments:
*   filename: name of file to save
*   image: Image struct to save
*
* Returns:
*   error code, 0 for success, 1 for failure
*/

int save_ppm_image_8(char* filename, RGBImage8* image) {

    FILE* f;
    int n_pixels;

    // Open the file
    f = fopen(filename, "wb");
    if (f == NULL) {
        fprintf(stderr, "Error opening file %s for write.\n", filename);
        return 1;
    }

    // Write the ppm header
    fprintf(f, "P6 %d %d %hu\n", image->width, image->height, 255);

    n_pixels = (image->width)*(image->height);

    // Write planar data in interleaved format
    for(int i=0; i < n_pixels; i++) {

        // Write pixel i: r, then g, then b, 1 record at a time
        fwrite(&image->data[0*n_pixels + i], sizeof(unsigned char), 1, f);
        fwrite(&image->data[1*n_pixels + i], sizeof(unsigned char), 1, f);
        fwrite(&image->data[2*n_pixels + i], sizeof(unsigned char), 1, f);
    }


    fclose(f);

    return 0;
}

File: projects/proj01_sol/code/scripts/pam_io.py
# Author: Stephen Timmel
# stimmel@vt.edu
# Not for external distribution.

#internal image layout: img[row][column][rgb]
#sample usage: img[5][5]['r']

import struct

def save_ppm_image_8(filename,img):
    f = open(filename,'w');

    #write ppm header
    width = len(img[0])
    height = len(img)
    f.write('P6 ' + str(width) + ' ' + str(height) + ' ' + '255' + '\n')

    for row in img:
        for pixel in row:
            red = struct.pack('@B',pixel['r'])
            green = struct.pack('@B',pixel['g'])
            blue = struct.pack('@B',pixel['b'])
            f.write(red + green + blue)
    f.close()

def load_ppm_image_8(filename):
    f = open(filename,'rb');

    #read ppm header
    header = f.readline().replace('\n','').split(' ')

    #validate header
    assert len(header) == 4, "Incorrect PPM8 header length"              
    assert header[0] == 'P6', "Incorrect PPM8 file format.  Must be P6"  
    assert header[1].isdigit(), "Incorrect PPM8 width.  Must be numeric" 
    assert header[2].isdigit(), "Incorrect PPM8 height. Must be numeric" 
    assert header[3].isdigit(), "Incorrect PPM8 color depth. Must be numeric"
    assert header[3] == '255', "Incorrect PPM8 color depth.  Must be 255"

    #save width and height for later
    width = int(header[1])
    height = int(header[2])
    
    #pull file into an array, three bytes at a time
    img = []
    for h in range(height):
        img.append([])
        for w in range(width):
            img[h].append({})
            data = f.read(3)
            img[h][w]['r'] = int(struct.unpack_from('@B',data,0)[0])
            img[h][w]['g'] = int(struct.unpack_from('@B',data,1)[0])
            img[h][w]['b'] = int(struct.unpack_from('@B',data,2)[0])

    f.close()
    return img

def save_pgm_image_8(filename,img):
    f = open(filename,'w');

    #write ppm header
    width = len(img[0])
    height = len(img)
    f.write('P5 ' + str(width) + ' ' + str(height) + ' ' + '255' + '\n')

    for row in img:
        for pixel in row:
            shade = struct.pack('@B',pixel)
            f.write(shade)
    f.close()

def load_pgm_image_8(filename):
    f = open(filename,'rb');

    #read ppm header
    header = f.readline().replace('\n','').split(' ')

    #validate header
    assert len(header) == 4, "Incorrect PGM8 header length"              
    assert header[0] == 'P5', "Incorrect PGM8 file format.  Must be P5"  
    assert header[1].isdigit(), "Incorrect PGM8 width.  Must be numeric" 
    assert header[2].isdigit(), "Incorrect PGM8 height. Must be numeric" 
    assert header[3].isdigit(), "Incorrect PGM8 color depth. Must be numeric"
    assert header[3] == '255', "Incorrect PGM8 color depth.  Must be 255"

    #save width and height for later
    width = int(header[1])
    height = int(header[2])
    
    #pull file into an array, three bytes at a time
    img = []
    for h in range(height):
        img.append([])
        for w in range(width):
            data = f.read(1)
            img[h].append(int(struct.unpack('@B',data)[0]))

    f.close()
    return img
